mender-image-prepare.sh

```diff
@@ -0,0 +1,535 @@
+#!/bin/bash
+
+show_help() {
+cat << EOF
+Usage: $0 [prepare_image | make_sdimg | make_sdcard] [-i|--image | -d|--device -r|--rootsize]
+
+        device - can be /dev/sd* or /dev/mmcblk*
+
+        prepare_image - shrinks existing embedded image
+                        obligatory parameters: --image
+
+        make_sdcard   - formats, partitions and sets the sd card
+                        obligatory parameters: --image & --rootsize
+                        
+                        note: rootsize can found as an output from
+                        'prepare_image' command
+
+        make_sdimg    - composes image file compliant with Mender partition
+                        table scheme
+
+EOF
+}
+
+if [ $# -eq 0 ]; then
+  show_help
+  exit 1
+fi
+
+if ! id | grep -q root; then
+        echo "must be run as root"
+        exit
+fi
+
+BASE_DIR=${PWD}
+
+# Erase block size 8MiB
+erase_block=8388608
+# Boot partition start in sectors (512 bytes per sector).
+pboot_start=16384
```

Comment:
I would expect this value to be parsed from the original image content. Generally I believe the goal should be to have a minimal set of deviations from the original image.

Reply:
Original images have 24MB for BBB and 12MB for RP PI3 of unallocated space before boot partition. The main requirement is to provide 8MB (equal to erase sector size) alignment. This is why I decided to have just 8MB unallocated space. If there is some crucial reason why we should stick with mentioned spaces, then of course I will change it. Kristian? Mirza?

Reply:
I think @mirzak is right. Originally I was thinking that we should make this match the Yocto build, but since we are going to use GRUB instead of patching U-Boot, it makes more sense to keep what's in the original image (in case the bootloader has some hardcoded offsets in it). Then our GRUB environment data will live inside the boot partition instead.

Reply:
My only motivation is to stay as close as possible to the original images, to avoid breaking anything that the images might contain. Example would be some dumb script that assumes a certain offset for some parsing and if we change the offset we break the script. I do not know if this "dumb script" exists but all deviations from the original image have the potential to break something.

It could also be that the boot ROM expects the boot-part to be at a certain hard-coded offset for the board to start.

Reply:
One question: does "original image" refer to Yocto build images or existing embedded images (Debian, Raspbian, etc.)?

So far this 8MiB unallocated space before boot partition comes from alignment used in Yocto builds. The same refers to all partitions sizes - they are all aligned to 8MiB.

In case of existing image for BBB there is only a single partition with 4MiB offset purposed for DOS/MBR boot sector.

For Raspbian there are two partitions: boot & rootfs. Also here an unallocated space before boot partition is 4MiB.

So please let me know whether I should track the unallocated space before boot partition individually for each existing embedded image or I can stay with 8MiB?

Reply:
> One question: does "original image" refer to Yocto build images or existing embedded images (Debian, Raspbian, etc.)?

Sorry for not being clear. 

"original image" == Debian, Raspbian etc.

But also:

"original image" == "input image"

It should not really matter where the input image is from, our changes should be minimal.

> So please let me know whether I should track the unallocated space before boot partition individually for each existing embedded image or I can stay with 8MiB?

I would suggest "individually for each existing embedded", preferable extracted from input image and note open-coded.